src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs

```diff
@@ -21,62 +24,127 @@ namespace Microsoft.CodeAnalysis.Completion.Providers
     /// </summary>
     internal abstract class AbstractAwaitCompletionProvider : LSPCompletionProvider
     {
+        #region CompletionItem properties keys
+        private const string AwaitCompletionTargetTokenPosition = nameof(AwaitCompletionTargetTokenPosition);
+
+        private static class AwaitCompletionChange
+        {
+            public const string AddAwaitAtCursor = nameof(AwaitCompletionChange) + "." + nameof(AddAwaitAtCursor);
+            public const string AddAwaitBeforeDotExpression = nameof(AwaitCompletionChange) + "." + nameof(AddAwaitBeforeDotExpression);
+            public const string AppendConfigureAwait = nameof(AwaitCompletionChange) + "." + nameof(AppendConfigureAwait);
+            public const string MakeContainerAsync = nameof(AwaitCompletionChange) + "." + nameof(MakeContainerAsync);
+        }
+        #endregion
+
+        protected enum DotAwaitContext
+        {
+            None,
+            AwaitOnly,
+            AwaitAndConfigureAwait,
+        }
+
         /// <summary>
         /// Gets the span start where async keyword should go.
         /// </summary>
         private protected abstract int GetSpanStart(SyntaxNode declaration);
 
         private protected abstract SyntaxNode? GetAsyncSupportingDeclaration(SyntaxToken token);
 
+        private protected abstract ITypeSymbol? GetTypeSymbolOfExpression(SemanticModel semanticModel, SyntaxNode potentialAwaitableExpression, CancellationToken cancellationToken);
+        private protected abstract SyntaxNode? GetExpressionToPlaceAwaitInFrontOf(SyntaxTree syntaxTree, int position, CancellationToken cancellationToken);
+        private protected abstract SyntaxToken? GetDotTokenLeftOfPosition(SyntaxTree syntaxTree, int position, CancellationToken cancellationToken);
+
+        private protected static bool IsConfigureAwaitable(Compilation compilation, ITypeSymbol symbol)
+        {
+            var originalDefinition = symbol.OriginalDefinition;
+            return
+                originalDefinition.Equals(compilation.TaskOfTType()) ||
+                originalDefinition.Equals(compilation.TaskType()) ||
+                originalDefinition.Equals(compilation.ValueTaskOfTType()) ||
+                originalDefinition.Equals(compilation.ValueTaskType());
+        }
```

Comment:
This is the place where we decide, if we offer "awaitf" or not. At the moment, the check is conservative and only offers on `Task`, `Task<T>`, `ValueTask`, and `ValueTask<T>`.  But we could:
* Offer if there is a ConfigureAwaitable (instance or extension method) with the fitting signature in scope (returns an Awaitable type and takes a boolean parameter). This could be done analog to what is done by [ConfigureAwaitChecker.Analyzer](https://github.com/cincuranet/ConfigureAwaitChecker/blob/11bd8923c80b654f4828338da2828c8ae9eb46b1/ConfigureAwaitChecker.Lib/Checker.cs#L230-L250)
* Offer always

But there are at least two cases, where we should not offer awaitf (I don't have tests for the cases yet):
```CS
async Task M()
{
  someTask.$$ConfigureAwait(false);
  someTask.ConfigureAwait(false).$$;
}
```

Reply:
tbh, i would just always offer ConfigureAwait(False).  Yes, tehre might be some cases where it's not legal, but tha'ts ok.  Completion is ok showing *too much* versus *too little*.  I'd prefer to remove this to reduce complecity.

Reply:
I would at least exclude `someTask.ConfigureAwait(false).$$`.

And if we ever extend this to support `IAsyncDisposable` we should not offer "ConfigureAwait" because of:
```CS
using (new MyAsyncDisposableClass().$$)

// awaitf would need to become

var c = new MyAsyncDisposableClass();
await using (c.ConfigureAwait(false))
{
    ...
}
```